syntax = "proto3";

package stt;

// STT Service for real-time speech-to-text transcription
service STTService {
  // Transcribe audio data to text with timing segments
  rpc Transcribe (AudioRequest) returns (TranscriptionResponse);

  // Stream audio chunks for real-time transcription
  rpc TranscribeStream (stream AudioChunk) returns (stream TranscriptionResponse);

  // Health check for service availability
  rpc HealthCheck (HealthCheckRequest) returns (HealthCheckResponse);
}

// Request message for single audio transcription
message AudioRequest {
  // Raw audio data (PCM format)
  bytes audio_data = 1;

  // Audio sample rate (e.g., 16000 for 16kHz)
  int32 sample_rate = 2;

  // Target language code (e.g., "it", "en"). Empty for auto-detection
  string language = 3;

  // Task type: "transcribe" or "translate"
  string task = 4;

  // Optional: Audio format metadata
  AudioFormat format = 5;

  // Optional: Request ID for tracking
  string request_id = 6;
}

// Audio format metadata
message AudioFormat {
  // Number of channels (1 = mono, 2 = stereo)
  int32 channels = 1;

  // Sample width in bytes (2 for 16-bit, 4 for 32-bit)
  int32 sample_width = 2;

  // Encoding (e.g., "pcm", "float32")
  string encoding = 3;
}

// Audio chunk for streaming transcription
message AudioChunk {
  // Audio data chunk
  bytes audio_data = 1;

  // Sequence number for ordering
  int32 sequence = 2;

  // True if this is the last chunk
  bool is_last = 3;

  // Optional: Request ID for tracking
  string request_id = 4;
}

// Response message with transcription result
message TranscriptionResponse {
  // Full transcription text
  string text = 1;

  // Transcription segments with timing
  repeated Segment segments = 2;

  // Detected/requested language code
  string language = 3;

  // Audio duration in seconds
  float duration = 4;

  // Processing time in milliseconds
  float processing_time_ms = 5;

  // Request ID (echo from request)
  string request_id = 6;

  // Status information
  Status status = 7;
}

// Individual transcription segment with timing
message Segment {
  // Segment text
  string text = 1;

  // Start time in seconds
  float start = 2;

  // End time in seconds
  float end = 3;

  // Confidence score (0.0 to 1.0)
  float confidence = 4;

  // Word-level timestamps
  repeated Word words = 5;
}

// Word-level timestamp information
message Word {
  // Word text
  string word = 1;

  // Start time in seconds
  float start = 2;

  // End time in seconds
  float end = 3;

  // Confidence score
  float confidence = 4;
}

// Status information
message Status {
  // Status code (0 = success, non-zero = error)
  int32 code = 1;

  // Human-readable message
  string message = 2;

  // Error details if applicable
  string error_details = 3;
}

// Health check request
message HealthCheckRequest {
  // Service name to check
  string service = 1;
}

// Health check response
message HealthCheckResponse {
  // Service status
  enum ServingStatus {
    UNKNOWN = 0;
    SERVING = 1;
    NOT_SERVING = 2;
    SERVICE_UNKNOWN = 3;
  }

  ServingStatus status = 1;

  // Additional health information
  string message = 2;

  // Model information
  ModelInfo model_info = 3;
}

// Model information for health checks
message ModelInfo {
  // Model name (e.g., "large-v3")
  string name = 1;

  // Device (e.g., "cuda", "cpu")
  string device = 2;

  // Compute type (e.g., "float16", "int8")
  string compute_type = 3;

  // Is model loaded
  bool is_loaded = 4;
}
